plugins/org.eclipse.n4js/src/org/eclipse/n4js/utils/ProjectDiscoveryHelper.java

```diff
@@ -215,4 +234,111 @@ public FileVisitResult visitFileFailed(Path file, IOException exc) {
 		}
 	}
 
+	/** @return the potentially cached {@link ProjectDescription} for the project in the given directory */
+	private ProjectDescription getCachedProjectDescription(Path path, Map<Path, ProjectDescription> cache) {
+		if (!cache.containsKey(path)) {
+			FileURI uri = new FileURI(path.toFile());
+			ProjectDescription depPD = projectDescriptionLoader.loadProjectDescriptionAtLocation(uri);
+			cache.put(path, depPD);
+		}
+
+		return cache.get(path);
+	}
+
+	/** Searches all dependencies (ie. npm projects) of the given set of projects */
+	@SuppressWarnings("unused")
+	private LinkedHashSet<Path> collectAllDependencies(LinkedHashSet<Path> allProjectDirs,
+			Map<Path, ProjectDescription> pdCache) {
+
+		LinkedHashSet<Path> dependencies = new LinkedHashSet<>();
+		List<Path> nodeModulesFolders = nodeModulesDiscoveryHelper.findNodeModulesFolders(allProjectDirs, pdCache);
+
+		for (Path nmFolder : new LinkedHashSet<>(nodeModulesFolders)) {
+			collectProjects(nmFolder, true, pdCache, dependencies);
+		}
+		return dependencies;
+	}
+
+	/**
+	 * This method will <b>not find all</b> dependencies but only those that are actually necessary for compiling all
+	 * the projects found in the given set of workspace directories.
+	 * <p>
+	 * Necessary project dependencies are those that are either N4JS projects or non-N4JS projects that are direct
+	 * dependencies of N4JS projects.
+	 */
+	private LinkedHashSet<Path> collectNecessaryDependencies(LinkedHashSet<Path> allProjectDirs,
```

Comment:
Since this behavior of omitting "unnecessary" dependencies can be a bit surprising, it would be good to document this as well in the API doc of the class `ProjectDiscoveryHelper` and on (some) public methods. Maybe have the main explanation in the class-level API doc and add links to that place in the doc of this method and (some) public methods.